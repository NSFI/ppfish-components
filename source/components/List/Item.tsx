import React from 'react';
import classNames from 'classnames';
import { Col } from '../Grid';
import { ListGridType, ColumnType } from './index';
import ListGridContext from './context';

export interface ListItemProps {
  className?: string;
  children?: React.ReactNode;
  prefixCls?: string;
  style?: React.CSSProperties;
  extra?: React.ReactNode;
  actions?: Array<React.ReactNode>;
  grid?: ListGridType;
}

export interface ListItemMetaProps {
  avatar?: React.ReactNode;
  className?: string;
  children?: React.ReactNode;
  description?: React.ReactNode;
  prefixCls?: string;
  style?: React.CSSProperties;
  title?: React.ReactNode;
}

export const Meta = (props: ListItemMetaProps) => {
  const { prefixCls = 'fishd-list', className, avatar, title, description, ...others } = props;

  const classString = classNames(`${prefixCls}-item-meta`, className);

  const content = (
    <div className={`${prefixCls}-item-meta-content`}>
      {title && <h4 className={`${prefixCls}-item-meta-title`}>{title}</h4>}
      {description && <div className={`${prefixCls}-item-meta-description`}>{description}</div>}
    </div>
  );

  return (
    <div {...others} className={classString}>
      {avatar && <div className={`${prefixCls}-item-meta-avatar`}>{avatar}</div>}
      {(title || description) && content}
    </div>
  );
};

function getGrid(grid: ListGridType, t: ColumnType) {
  return grid[t] && Math.floor(24 / grid[t]!);
}

export default class Item extends React.Component<ListItemProps, any> {
  static Meta: typeof Meta = Meta;

  static contextType = ListGridContext

  context!: React.ContextType<typeof ListGridContext>

  render() {
    const { grid } = this.context;
    const { prefixCls = 'fishd-list', children, actions, extra, className, ...others } = this.props;
    const classString = classNames(`${prefixCls}-item`, className);

    const metaContent: React.ReactElement<any>[] = [];

    const otherContent: React.ReactElement<any>[] = [];

    React.Children.forEach(children, (element: React.ReactElement<any>) => {
      if (element && element.type && element.type === Meta) {
        metaContent.push(element);
      } else {
        otherContent.push(element);
      }
    });

    const contentClassString = classNames(`${prefixCls}-item-content`, {
      [`${prefixCls}-item-content-single`]: metaContent.length < 1
    });
    const content =
      otherContent.length > 0 ? <div className={contentClassString}>{otherContent}</div> : null;

    let actionsContent;
    if (actions && actions.length > 0) {
      const actionsContentItem = (action: React.ReactNode, i: number) => (
        <li key={`${prefixCls}-item-action-${i}`}>
          {action}
          {i !== actions.length - 1 && <em className={`${prefixCls}-item-action-split`} />}
        </li>
      );
      actionsContent = (
        <ul className={`${prefixCls}-item-action`}>
          {actions.map((action, i) => actionsContentItem(action, i))}
        </ul>
      );
    }

    const extraContent = (
      <div className={`${prefixCls}-item-extra-wrap`}>
        <div className={`${prefixCls}-item-main`}>
          {metaContent}
          {content}
          {actionsContent}
        </div>
        <div className={`${prefixCls}-item-extra`}>{extra}</div>
      </div>
    );

    const mainContent = grid ? (
      <Col
        span={getGrid(grid, 'column')}
        xs={getGrid(grid, 'xs')}
        sm={getGrid(grid, 'sm')}
        md={getGrid(grid, 'md')}
        lg={getGrid(grid, 'lg')}
        xl={getGrid(grid, 'xl')}
        xxl={getGrid(grid, 'xxl')}
      >
        <div {...others} className={classString}>
          {extra && extraContent}
          {!extra && metaContent}
          {!extra && content}
          {!extra && actionsContent}
        </div>
      </Col>
    ) : (
      <div {...others} className={classString}>
        {extra && extraContent}
        {!extra && metaContent}
        {!extra && content}
        {!extra && actionsContent}
      </div>
    );

    return mainContent;
  }
}
