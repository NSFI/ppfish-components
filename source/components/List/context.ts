import { createContext } from 'react';
import { ListGridType } from './index';

const ListGridContext = createContext<{ grid: ListGridType }>({ grid: {} });

export const ListGridContextProvider = ListGridContext.Provider;

export const ListGridContextConsumer = ListGridContext.Consumer;

export default ListGridContext;
