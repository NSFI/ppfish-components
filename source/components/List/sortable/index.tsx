import React from 'react';
import { SortableComposition as _Sortable } from './SortableComposition';

export { HORIZONTAL } from './SortableComposition';
export { VERTICAL } from './SortableComposition';

function SortableItem(props: any) {
  return (
    <div {...props} className="fishd-list-sortable-item">
      {props.children}
    </div>
  );
}

const Sortable = _Sortable(SortableItem);

export default Sortable;
