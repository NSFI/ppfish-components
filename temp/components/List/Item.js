var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import * as React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Col } from '../Grid';
export var Meta = function (props) {
    var _a = props.prefixCls, prefixCls = _a === void 0 ? 'fishd-list' : _a, className = props.className, avatar = props.avatar, title = props.title, description = props.description, others = __rest(props, ["prefixCls", "className", "avatar", "title", "description"]);
    var classString = classNames(prefixCls + "-item-meta", className);
    var content = (React.createElement("div", { className: prefixCls + "-item-meta-content" },
        title && React.createElement("h4", { className: prefixCls + "-item-meta-title" }, title),
        description && React.createElement("div", { className: prefixCls + "-item-meta-description" }, description)));
    return (React.createElement("div", __assign({}, others, { className: classString }),
        avatar && React.createElement("div", { className: prefixCls + "-item-meta-avatar" }, avatar),
        (title || description) && content));
};
function getGrid(grid, t) {
    return grid[t] && Math.floor(24 / grid[t]);
}
var GridColumns = ['', 1, 2, 3, 4, 6, 8, 12, 24];
var Item = /** @class */ (function (_super) {
    __extends(Item, _super);
    function Item() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Item.prototype.render = function () {
        var _a;
        var grid = this.context.grid;
        var _b = this.props, _c = _b.prefixCls, prefixCls = _c === void 0 ? 'fishd-list' : _c, children = _b.children, actions = _b.actions, extra = _b.extra, className = _b.className, others = __rest(_b, ["prefixCls", "children", "actions", "extra", "className"]);
        var classString = classNames(prefixCls + "-item", className);
        var metaContent = [];
        var otherContent = [];
        React.Children.forEach(children, function (element) {
            if (element && element.type && element.type === Meta) {
                metaContent.push(element);
            }
            else {
                otherContent.push(element);
            }
        });
        var contentClassString = classNames(prefixCls + "-item-content", (_a = {},
            _a[prefixCls + "-item-content-single"] = metaContent.length < 1,
            _a));
        var content = otherContent.length > 0 ? React.createElement("div", { className: contentClassString }, otherContent) : null;
        var actionsContent;
        if (actions && actions.length > 0) {
            var actionsContentItem_1 = function (action, i) { return (React.createElement("li", { key: prefixCls + "-item-action-" + i },
                action,
                i !== actions.length - 1 && React.createElement("em", { className: prefixCls + "-item-action-split" }))); };
            actionsContent = (React.createElement("ul", { className: prefixCls + "-item-action" }, actions.map(function (action, i) { return actionsContentItem_1(action, i); })));
        }
        var extraContent = (React.createElement("div", { className: prefixCls + "-item-extra-wrap" },
            React.createElement("div", { className: prefixCls + "-item-main" },
                metaContent,
                content,
                actionsContent),
            React.createElement("div", { className: prefixCls + "-item-extra" }, extra)));
        var mainContent = grid ? (React.createElement(Col, { span: getGrid(grid, 'column'), xs: getGrid(grid, 'xs'), sm: getGrid(grid, 'sm'), md: getGrid(grid, 'md'), lg: getGrid(grid, 'lg'), xl: getGrid(grid, 'xl'), xxl: getGrid(grid, 'xxl') },
            React.createElement("div", __assign({}, others, { className: classString }),
                extra && extraContent,
                !extra && metaContent,
                !extra && content,
                !extra && actionsContent))) : (React.createElement("div", __assign({}, others, { className: classString }),
            extra && extraContent,
            !extra && metaContent,
            !extra && content,
            !extra && actionsContent));
        return mainContent;
    };
    Item.Meta = Meta;
    Item.propTypes = {
        column: PropTypes.oneOf(GridColumns),
        xs: PropTypes.oneOf(GridColumns),
        sm: PropTypes.oneOf(GridColumns),
        md: PropTypes.oneOf(GridColumns),
        lg: PropTypes.oneOf(GridColumns),
        xl: PropTypes.oneOf(GridColumns),
        xxl: PropTypes.oneOf(GridColumns)
    };
    Item.contextTypes = {
        grid: PropTypes.any
    };
    return Item;
}(React.Component));
export default Item;
