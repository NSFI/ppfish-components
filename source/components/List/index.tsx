import React, { ReactNode } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Spin, { SpinProps } from '../Spin';
import Sortable from './sortable';
import Pagination, { PaginationConfig } from '../Pagination';
import { Row } from '../Grid';

import Item from './Item';
import './style/index.less';
import ConfigConsumer from '../Config/Consumer';
import { LocaleProperties } from '../Locale';

export { ListItemProps, ListItemMetaProps } from './Item';

export type ColumnCount = 1 | 2 | 3 | 4 | 6 | 8 | 12 | 24;

export type ColumnType = 'gutter' | 'column' | 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl';

export interface ListGridType {
  gutter?: number;
  column?: ColumnCount;
  xs?: ColumnCount;
  sm?: ColumnCount;
  md?: ColumnCount;
  lg?: ColumnCount;
  xl?: ColumnCount;
  xxl?: ColumnCount;
}

export type ListSize = 'small' | 'default' | 'large';

export interface ListProps {
  bordered?: boolean;
  className?: string;
  children?: React.ReactNode;
  dataSource: any;
  extra?: React.ReactNode;
  grid?: ListGridType;
  id?: string;
  itemLayout?: string;
  loading?: boolean | SpinProps;
  loadMore?: React.ReactNode;
  pagination?: PaginationConfig;
  prefixCls?: string;
  rowKey?: any;
  renderItem: any;
  size?: ListSize;
  split?: boolean;
  header?: React.ReactNode;
  footer?: React.ReactNode;
  locale?: LocaleProperties['List'];
  striped?: Boolean;
}

class List extends React.Component<ListProps> {
  static Item: typeof Item = Item;
  static Sortable = Sortable;
  static childContextTypes = {
    grid: PropTypes.any,
  };

  static propTypes = {
    locale: PropTypes.shape({
      emptyText: PropTypes.string,
    }),
    dataSource: PropTypes.array,
    prefixCls: PropTypes.string,
    bordered: PropTypes.bool,
    split: PropTypes.bool,
    loading: PropTypes.bool,
    pagination: PropTypes.bool,
    striped: PropTypes.bool,
  };

  static defaultProps = {
    dataSource: [],
    prefixCls: 'fishd-list',
    bordered: false,
    split: true,
    loading: false,
    pagination: false,
    striped: false,
  };

  state = {
    paginationCurrent: 1,
  };

  defaultPaginationProps = {
    current: 1,
    pageSize: 10,
    onChange: (page: number, pageSize: number) => {
      const { pagination } = this.props;
      this.setState({
        paginationCurrent: page,
      });
      if (pagination && pagination.onChange) {
        pagination.onChange(page, pageSize);
      }
    },
    total: 0,
  };

  private keys: { [key: string]: string } = {};

  getChildContext() {
    return {
      grid: this.props.grid,
    };
  }

  renderItem = (item: React.ReactElement<any>, index: number) => {
    const { dataSource, renderItem, rowKey } = this.props;
    let key;

    if (typeof rowKey === 'function') {
      key = rowKey(dataSource[index]);
    } else if (typeof rowKey === 'string') {
      key = dataSource[rowKey];
    } else {
      key = dataSource.key;
    }

    if (!key) {
      key = `list-item-${index}`;
    }

    this.keys[index] = key;

    return renderItem(item, index);
  };

  isSomethingAfterLastItem() {
    const { loadMore, pagination, footer } = this.props;
    return !!(loadMore || pagination || footer);
  }

  renderEmpty = (Locale: LocaleProperties['List']) => {
    const defaultEmptyText = Locale ? Locale.emptyText : '暂无数据';
    const emptyText = this.props.locale ? this.props.locale.emptyText : defaultEmptyText;
    return <div className={`${this.props.prefixCls}-empty-text`}>{emptyText}</div>;
  };

  render() {
    const { paginationCurrent } = this.state;
    const {
      bordered,
      split,
      className,
      children,
      itemLayout,
      loadMore,
      pagination,
      prefixCls,
      grid,
      dataSource,
      size,
      rowKey,
      renderItem,
      header,
      footer,
      loading,
      striped,
      ...rest
    } = this.props;

    let loadingProp = loading;
    if (typeof loadingProp === 'boolean') {
      loadingProp = {
        spinning: loadingProp,
      };
    }
    const isLoading = loadingProp && loadingProp.spinning;

    // large => lg
    // small => sm
    let sizeCls = '';
    switch (size) {
      case 'large':
        sizeCls = 'lg';
        break;
      case 'small':
        sizeCls = 'sm';
        break;
    }

    const classString = classNames(prefixCls, className, {
      [`${prefixCls}-vertical`]: itemLayout === 'vertical',
      [`${prefixCls}-${sizeCls}`]: sizeCls,
      [`${prefixCls}-split`]: split,
      [`${prefixCls}-bordered`]: bordered,
      [`${prefixCls}-loading`]: isLoading,
      [`${prefixCls}-grid`]: grid,
      [`${prefixCls}-striped`]: striped,
      [`${prefixCls}-something-after-last-item`]: this.isSomethingAfterLastItem(),
    });

    const paginationProps = {
      ...this.defaultPaginationProps,
      total: dataSource.length,
      current: paginationCurrent,
      ...(pagination || {}),
    };

    const largestPage = Math.ceil(paginationProps.total / paginationProps.pageSize);
    if (paginationProps.current > largestPage) {
      paginationProps.current = largestPage;
    }
    const paginationContent = pagination ? (
      <div className={`${prefixCls}-pagination`}>
        <Pagination {...paginationProps} onChange={this.defaultPaginationProps.onChange} />
      </div>
    ) : null;

    let splitDataSource = [...dataSource];
    if (pagination) {
      if (dataSource.length > (paginationProps.current - 1) * paginationProps.pageSize) {
        splitDataSource = [...dataSource].splice(
          (paginationProps.current - 1) * paginationProps.pageSize,
          paginationProps.pageSize,
        );
      }
    }

    const renderChildrenContent = (Locale: LocaleProperties['List']) => {
      let childrenContent: ReactNode;
      childrenContent = isLoading && <div style={{ minHeight: 53 }} />;
      if (splitDataSource.length > 0) {
        const items = splitDataSource.map((item: any, index: number) =>
          this.renderItem(item, index),
        );
        const childrenList = React.Children.map(items, (child: any, index) =>
          React.cloneElement(child, {
            key: this.keys[index],
          }),
        );

        childrenContent = grid ? <Row gutter={grid.gutter}>{childrenList}</Row> : childrenList;
      } else if (!children && !isLoading) {
        childrenContent = <div>{this.renderEmpty(Locale)}</div>;
      }

      return childrenContent;
    };

    const paginationPosition = paginationProps.position || 'bottom';

    return (
      <ConfigConsumer componentName="List">
        {(Locale: LocaleProperties['List']) => {
          return (
            <div className={classString} {...rest}>
              {(paginationPosition === 'top' || paginationPosition === 'both') && paginationContent}
              {header && <div className={`${prefixCls}-header`}>{header}</div>}
              <Spin {...loadingProp}>
                {renderChildrenContent(Locale)}
                {children}
              </Spin>
              {footer && <div className={`${prefixCls}-footer`}>{footer}</div>}
              {loadMore ||
                ((paginationPosition === 'bottom' || paginationPosition === 'both') &&
                  paginationContent)}
            </div>
          );
        }}
      </ConfigConsumer>
    );
  }
}

export default List;
