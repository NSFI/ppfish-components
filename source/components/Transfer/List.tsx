import * as React from 'react';
import * as ReactDOM from 'react-dom';
import classNames from 'classnames';
import Checkbox from '../Checkbox';
import PureRenderMixin from '../Checkbox/src/PureRenderMixin';
import { TransferItem } from './index';
import Search from './Search';
import Item from './Item';
import Animate from 'rc-animate';
import triggerEvent from '../../utils/triggerEvent';

// case sensitive
function isIEorEDGE() {
  return (document as any).documentMode || /Edge/.test(navigator.userAgent);
}

function noop() {}

function isRenderResultPlainObject(result: any) {
  return (
    result &&
    !React.isValidElement(result) &&
    Object.prototype.toString.call(result) === '[object Object]'
  );
}

export interface TransferListProps {
  mode?: 'single' | 'multiple';
  direction: string;
  prefixCls: string;
  titleText: string;
  dataSource: TransferItem[];
  filter: string;
  filterOption?: (filterText: any, item: any) => boolean;
  style?: React.CSSProperties;
  checkedKeys: string[];
  handleFilter: (e: any) => void;
  handleSelect: (selectedItem: any, checked: boolean) => void;
  handleSelectAll: (dataSource: any[], checkAll: boolean) => void;
  handleClose?: (selectedItem: any) => void;
  handleClear: () => void;
  render?: (item: any) => any;
  showSearch?: boolean;
  searchPlaceholder: string;
  notFoundContent: React.ReactNode;
  itemUnit: string;
  itemsUnit: string;
  body?: (props: any) => any;
  footer?: (props: any) => any;
  lazy?: boolean | {};
  onScroll: Function;
}

export default class TransferList extends React.Component<TransferListProps, any> {
  static defaultProps = {
    dataSource: [],
    titleText: '',
    showSearch: false,
    render: noop,
    lazy: {}
  };

  timer: number;
  triggerScrollTimer: number;
  fixIERepaintTimer: number;
  notFoundNode: HTMLDivElement;

  constructor(props: TransferListProps) {
    super(props);
    this.state = {
      mounted: false
    };
  }

  componentDidMount() {
    this.timer = window.setTimeout(() => {
      this.setState({
        mounted: true
      });
    }, 0);
  }

  componentWillUnmount() {
    clearTimeout(this.timer);
    clearTimeout(this.triggerScrollTimer);
    clearTimeout(this.fixIERepaintTimer);
  }

  shouldComponentUpdate(...args: any[]) {
    return PureRenderMixin.shouldComponentUpdate.apply(this, args);
  }

  getCheckStatus(filteredDataSource: TransferItem[]) {
    const { checkedKeys } = this.props;
    if (checkedKeys.length === 0) {
      return 'none';
    } else if (filteredDataSource.every(item => checkedKeys.indexOf(item.key) >= 0)) {
      return 'all';
    }
    return 'part';
  }

  handleSelect = (selectedItem: TransferItem, direction) => {
    const { checkedKeys, mode } = this.props;
    if (mode === 'single') {
      if (direction === 'right') {
        return;
      }
      this.props.handleSelect(selectedItem, true);
    } else {
      const result = checkedKeys.some(key => key === selectedItem.key);
      this.props.handleSelect(selectedItem, !result);
    }
  };

  // single模式下，点击目标列表中的关闭按钮
  handleClose = (selectedItem: TransferItem) => {
    this.props.handleClose(selectedItem);
  };

  handleFilter = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.props.handleFilter(e);
    if (!e.target.value) {
      return;
    }
    // Manually trigger scroll event for lazy search bug
    // https://github.com/ant-design/ant-design/issues/5631
    this.triggerScrollTimer = window.setTimeout(() => {
      const transferNode = ReactDOM.findDOMNode(this) as Element;
      const listNode = transferNode.querySelectorAll('.fishd-transfer-list-content')[0];
      if (listNode) {
        triggerEvent(listNode, 'scroll');
      }
    }, 0);
    this.fixIERepaint();
  };

  handleClear = () => {
    this.props.handleClear();
    this.fixIERepaint();
  };

  matchFilter = (text: string, item: TransferItem) => {
    const { filter, filterOption } = this.props;
    if (filterOption) {
      return filterOption(filter, item);
    }
    return text.indexOf(filter) >= 0;
  };

  renderItem = (item: TransferItem) => {
    const { render = noop } = this.props;
    const renderResult = render(item);
    const isRenderResultPlain = isRenderResultPlainObject(renderResult);
    return {
      renderedText: isRenderResultPlain ? renderResult.value : renderResult,
      renderedEl: isRenderResultPlain ? renderResult.label : renderResult
    };
  };

  saveNotFoundRef = (node: HTMLDivElement) => {
    this.notFoundNode = node;
  };

  // Fix IE/Edge repaint
  // https://github.com/ant-design/ant-design/issues/9697
  // https://stackoverflow.com/q/27947912/3040605
  fixIERepaint() {
    if (!isIEorEDGE()) {
      return;
    }
    this.fixIERepaintTimer = window.setTimeout(() => {
      if (this.notFoundNode) {
        // eslint-disable-next-line no-self-assign
        this.notFoundNode.className = this.notFoundNode.className;
      }
    }, 0);
  }

  render() {
    const {
      mode,
      direction,
      prefixCls,
      dataSource,
      titleText,
      checkedKeys,
      lazy,
      body = noop,
      footer = noop,
      showSearch,
      style,
      filter,
      searchPlaceholder,
      notFoundContent,
      itemUnit,
      itemsUnit,
      onScroll
    } = this.props;

    // Custom Layout
    const footerDom = footer({ ...this.props });
    const bodyDom = body({ ...this.props });

    const listCls = classNames(prefixCls, {
      [`${prefixCls}-with-footer`]: !!footerDom
    });

    const filteredDataSource: TransferItem[] = [];
    const totalDataSource: TransferItem[] = [];

    const showItems = dataSource.map(item => {
      const { renderedText, renderedEl } = this.renderItem(item);
      if (filter && filter.trim() && !this.matchFilter(renderedText, item)) {
        return null;
      }

      // all show items
      totalDataSource.push(item);
      if (!item.disabled) {
        // response to checkAll items
        filteredDataSource.push(item);
      }

      const checked = checkedKeys.indexOf(item.key) >= 0;
      return (
        <Item
          mode={mode}
          direction={direction}
          key={item.key}
          item={item}
          lazy={lazy}
          renderedText={renderedText}
          renderedEl={renderedEl}
          checked={checked}
          prefixCls={prefixCls}
          onClick={this.handleSelect}
          onClose={this.handleClose}
        />
      );
    });

    const unit = dataSource.length > 1 ? itemsUnit : itemUnit;

    const search = showSearch ? (
      <div className={`${prefixCls}-body-search-wrapper`}>
        <Search
          prefixCls={`${prefixCls}-search`}
          onChange={this.handleFilter}
          handleClear={this.handleClear}
          placeholder={searchPlaceholder}
          value={filter}
        />
      </div>
    ) : null;

    const listBody = bodyDom || (
      <div
        className={
          showSearch ? `${prefixCls}-body ${prefixCls}-body-with-search` : `${prefixCls}-body`
        }
      >
        {search}
        <Animate
          component="ul"
          componentProps={{ onScroll }}
          className={`${prefixCls}-content`}
          transitionName={this.state.mounted ? `${prefixCls}-content-item-highlight` : ''}
          transitionLeave={false}
        >
          {showItems}
        </Animate>
        <div className={`${prefixCls}-body-not-found`} ref={this.saveNotFoundRef}>
          {notFoundContent}
        </div>
      </div>
    );

    const listFooter = footerDom ? <div className={`${prefixCls}-footer`}>{footerDom}</div> : null;

    const checkStatus = this.getCheckStatus(filteredDataSource);
    const checkedAll = checkStatus === 'all';
    const checkAllCheckbox = (
      <Checkbox
        checked={checkedAll}
        indeterminate={checkStatus === 'part'}
        onChange={() => this.props.handleSelectAll(filteredDataSource, checkedAll)}
      />
    );

    return (
      <div className={listCls} style={style}>
        <div className={`${prefixCls}-header`}>
          {mode === 'multiple' ? checkAllCheckbox : null}
          <span className={`${prefixCls}-header-title`}>{titleText}</span>
          <span className={`${prefixCls}-header-selected`}>
            {mode === 'multiple'
              ? `${checkedKeys.length}/${totalDataSource.length}`
              : `${totalDataSource.length}`}{' '}
            {unit}
          </span>
        </div>
        {listBody}
        {listFooter}
      </div>
    );
  }
}
