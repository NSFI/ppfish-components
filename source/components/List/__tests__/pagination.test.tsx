import React from 'react';
import {render, mount} from 'enzyme';
import List from '../index';

describe('List.pagination', () => {
  const data = [
    {key: 0, name: 'Jack'},
    {key: 1, name: 'Lucy'},
    {key: 2, name: 'Tom'},
    {key: 3, name: 'Jerry'},
  ];

  const pagination = {className: 'my-page', pageSize: 2};

  function createList(props) {
    return (
      <List
        itemLayout="vertical"
        pagination={pagination}
        dataSource={data}
        renderItem={item => (
          <List.Item key={item.key}>
            {item.name}
          </List.Item>
        )}
        {...props}
      />
    );
  }

  function renderedNames(wrapper) {
    return wrapper.find('.fishd-list-item-content').map(row => row.text());
  }

  it('renders pagination correctly', () => {
    const wrapper = render(createList({}));
    expect(wrapper).toMatchSnapshot();
  });

  it('should not show pager if pagination.hideOnSinglePage is true and only 1 page', () => {
    const wrapper = mount(
      createList({pagination: {pageSize: 3, hideOnSinglePage: true}})
    );
    expect(wrapper.find('.fishd-pagination')).toHaveLength(1);
    wrapper.setProps({pagination: {pageSize: 3, hideOnSinglePage: false}});
    expect(wrapper.find('.fishd-pagination')).toHaveLength(1);
    wrapper.setProps({pagination: {pageSize: 4, hideOnSinglePage: true}});
    expect(wrapper.find('.fishd-pagination')).toHaveLength(0);
    wrapper.setProps({pagination: {pageSize: 4, hideOnSinglePage: false}});
    expect(wrapper.find('.fishd-pagination')).toHaveLength(1);
    wrapper.setProps({pagination: {pageSize: 5, hideOnSinglePage: true}});
    expect(wrapper.find('.fishd-pagination')).toHaveLength(0);
    wrapper.setProps({pagination: {pageSize: 5, hideOnSinglePage: false}});
    expect(wrapper.find('.fishd-pagination')).toHaveLength(1);
  });

  it('paginate data', () => {
    const wrapper = mount(createList({}));

    expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy']);
    wrapper
      .find('Pager')
      .last()
      .simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Jerry']);
  });

  it('repaginates when pageSize change', () => {
    const wrapper = mount(createList({}));

    wrapper.setProps({pagination: {pageSize: 1}});
    expect(renderedNames(wrapper)).toEqual(['Jack']);
  });

  it('fires change event', () => {
    const handlePaginationChange = jest.fn();
    const noop = () => {
    };
    const wrapper = mount(
      createList({
        pagination: {
          ...pagination,
          onChange: handlePaginationChange,
          onShowSizeChange: noop,
        },
      })
    );

    wrapper
      .find('Pager')
      .last()
      .simulate('click');

    expect(handlePaginationChange).toBeCalledWith(2, 2);
  });

  // https://codepen.io/afc163/pen/pWVRJV?editors=001
  it('should display pagination as prop pagination change between true and false', () => {
    const wrapper = mount(createList({}));
    expect(wrapper.find('.fishd-pagination')).toHaveLength(1);
    expect(wrapper.find('.fishd-pagination-item')).toHaveLength(2);
    wrapper.setProps({pagination: false});
    expect(wrapper.find('.fishd-pagination')).toHaveLength(0);
    wrapper.setProps({pagination});
    wrapper.update();
    expect(wrapper.find('.fishd-pagination')).toHaveLength(1);
    expect(wrapper.find('.fishd-pagination-item')).toHaveLength(2);
    wrapper.find('.fishd-pagination-item-2').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Jerry']);
    wrapper.setProps({pagination: false});
    expect(wrapper.find('.fishd-pagination')).toHaveLength(0);
    wrapper.setProps({pagination: true});
    expect(wrapper.find('.fishd-pagination')).toHaveLength(1);
    expect(wrapper.find('.fishd-pagination-item')).toHaveLength(1); // pageSize will be 10
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy', 'Tom', 'Jerry']);
  });

  it('change to correct page when data source changes', () => {
    const wrapper = mount(createList({pagination: {pageSize: 1}}));
    wrapper.find('.fishd-pagination-item-3').simulate('click');
    wrapper.setProps({dataSource: [data[0]]});
    expect(
      wrapper
        .find('.fishd-pagination-item-1')
        .hasClass('fishd-pagination-item-active')
    ).toBe(true);
  });

  it('specify the position of pagination', () => {
    const wrapper = mount(createList({pagination: {position: 'top'}}));
    expect(wrapper.find('.fishd-list').childAt(0).find('.fishd-pagination')).toHaveLength(1);
    wrapper.setProps({pagination: {position: 'bottom'}});
    expect(wrapper.find('.fishd-list').children().last().find('.fishd-pagination')).toHaveLength(1);
    wrapper.setProps({pagination: {position: 'both'}});
    expect(wrapper.find('.fishd-pagination')).toHaveLength(2);
    expect(wrapper.find('.fishd-list').childAt(0).find('.fishd-pagination')).toHaveLength(1);
    expect(wrapper.find('.fishd-list').children().last().find('.fishd-pagination')).toHaveLength(1);
  });
});
