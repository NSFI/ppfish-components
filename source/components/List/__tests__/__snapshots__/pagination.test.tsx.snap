// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List.pagination renders pagination correctly 1`] = `
<div
  class="fishd-list fishd-list-vertical fishd-list-split fishd-list-something-after-last-item"
>
  <div
    class="fishd-spin-nested-loading"
  >
    <div
      class="fishd-spin-nested"
    >
      <div
        class="fishd-list-item"
      >
        <div
          class="fishd-list-item-content fishd-list-item-content-single"
        >
          Jack
        </div>
      </div>
      <div
        class="fishd-list-item"
      >
        <div
          class="fishd-list-item-content fishd-list-item-content-single"
        >
          Lucy
        </div>
      </div>
    </div>
  </div>
  <div
    class="fishd-list-pagination"
  >
    <ul
      class="fishd-pagination my-page"
      unselectable="unselectable"
    >
      <li
        aria-disabled="true"
        class="fishd-pagination-disabled fishd-pagination-prev"
        title="上一页"
      >
        <a
          class="fishd-pagination-item-link"
        />
      </li>
      <li
        class="fishd-pagination-item fishd-pagination-item-1 fishd-pagination-item-active"
        tabindex="0"
        title="1"
      >
        <a>
          1
        </a>
      </li>
      <li
        class="fishd-pagination-item fishd-pagination-item-2"
        tabindex="0"
        title="2"
      >
        <a>
          2
        </a>
      </li>
      <li
        aria-disabled="false"
        class=" fishd-pagination-next"
        tabindex="0"
        title="下一页"
      >
        <a
          class="fishd-pagination-item-link"
        />
      </li>
    </ul>
  </div>
</div>
`;
